File: ruby/leetcode/kth-smallest-element-in-a-bst.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @param {Integer} k
# @return {Integer}
def kth_smallest(root, k)
    def traverse(root)
        if root == nil
            return []
        end
        return traverse(root.left) + [root.val] + traverse(root.right)
    end
    return traverse(root)[k - 1]
end

File: javascript/leetcode/kth-smallest-element-in-a-bst.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} k
 * @return {number}
 */
var kthSmallest = function(root, k) {
    const traverse = function(root) {
        if (root == null) { return [] }
        return traverse(root.left).concat([root.val]).concat(traverse(root.right));
    }
    return traverse(root)[k - 1];
};
